File: cocos2dx-js/Cocos2dxNode.js
// var cc = cc

var AbstractNode = hunter.require('support.poco.sdk.AbstractNode')

var Node = function (node, screenWidth, screenHeight) {
    AbstractNode.call(this)
    this.node = node
    this.screenWidth = screenWidth
    this.screenHeight = screenHeight
}
Node.prototype = Object.create(AbstractNode.prototype)

Node.prototype.getParent = function () {
    var parent = this.node.getParent()
    if (!parent) {
        return null
    }
    return new Node(parent, this.screenWidth, this.screenHeight)
}

Node.prototype.getChildren = function () {
    var children = null
    var nodeChildren = this.node.getChildren()
    if (nodeChildren) {
        children = []
        for (var i in nodeChildren) {
            var child = nodeChildren[i]
            children.push(new Node(child, this.screenWidth, this.screenHeight))
        }
    }
    return children
}

Node.prototype.getAttr = function (attrName) {
    if (attrName === 'visible') {
        if (this.node.isVisible) {
            var visible = this.node.isVisible()
            if (!visible) {
                return false
            }

            // if the node is visible, check its parent's visibility
            // 举一个很简单的例子：
            // 有两个layer，一个layer中有个button，这个button点下去后把layer的visible设为false，
            // 这时候这个button的visible仍然是true的，如果这时候判断这个button是否存在或可见，
            // 则需要判断他的所有父节点是否可见了
            var parent = this.node.getParent()
            while (parent) {
                var parentVisible = parent.isVisible()
                if (!parentVisible) {
                    return false
                }
                parent = parent.getParent()
            }
            return true
        } else {
            return this.node._activeInHierarchy
        }
    }
    else if (attrName === 'name') {
        return this.node.getName() || '<no-name>'
    }
    else if (attrName === 'text') {
        for (var i in this.node._components) {
            var c = this.node._components[i]
            if (c.string !== undefined) {
                return c.string
            }
        }
    }
    else if (attrName === 'type') {
        var ntype = ''
        if (this.node._components) {
            for (var i = this.node._components.length - 1; i >= 0; i--) {
                ntype = this.node._components[i].__classname__
                if (ntype.startsWith('cc')) {
                    break
                }
            }
        } else {
            ntype = this.node.__classname__
        }
        return ntype.replace(/\w+\./, '')
    }
    else if (attrName === 'pos') {
        // 转换成归一化坐标系，原点左上角
        var pos = this.node.convertToWorldSpaceAR(new cc.Vec2(0, 0))
        pos.x /= this.screenWidth
        pos.y /= this.screenHeight
        pos.y = 1 - pos.y
        return [pos.x, pos.y]
    }
    else if (attrName === 'size') {
        // 转换成归一化坐标系
        var size = new cc.Size(this.node.getContentSize())
        size.width /= this.screenWidth
        size.height /= this.screenHeight
        return [size.width, size.height]
    }
    else if (attrName === 'scale') {
        return [this.node.getScaleX(), this.node.getScaleY()]
    }
    else if (attrName === 'anchorPoint') {
        var anchor = this.node.getAnchorPoint()
        return [anchor.x, 1 - anchor.y]
    }
    else if (attrName === 'zOrders') {
        return {
            local: this.node.getLocalZOrder(), 
            global: this.node.getGlobalZOrder(),
        }
    }
    else if (attrName == 'touchable') {
        if (this.node.isTouchEnabled) {
            return this.node.isTouchEnabled()
        }
    }
    else if (attrName === 'tag') {
        return this.node.getTag()
    }
    else if (attrName === 'enabled') {
        if (this.node.isEnabled) {
            return this.node.isEnabled()
        }
    }
    else if (attrName === 'rotation') {
        return this.node.getRotation()
    }

    return undefined
}

Node.prototype.setAttr = function (attrName, val) {
    // raise UnableToSetAttributeException(attrName, self.node)
}

Node.prototype.getAvailableAttributeNames = function () {
    // enumerate all available attributes' name of this node
    // :rettype: Iterable<string>

    return AbstractNode.prototype.getAvailableAttributeNames.call(this).concat([
        'text',
        'touchable',
        'enabled',
        'tag',
        'rotation',
    ])
}

module.exports = Node;

File: cocos2dx-lua/poco/Cocos2dxNode.lua
-- import
function import(moduleName, currentModuleName)
    local currentModuleNameParts
    local moduleFullName = moduleName
    local offset = 1

    while true do
        if string.byte(moduleName, offset) ~= 46 then
            moduleFullName = string.sub(moduleName, offset)
            if currentModuleNameParts and #currentModuleNameParts > 0 then
                moduleFullName = table.concat(currentModuleNameParts, ".") .. "." .. moduleFullName
            end
            break
        end
        offset = offset + 1

        if not currentModuleNameParts then
            if not currentModuleName then
                local n, v = debug.getlocal(3, 1)
                currentModuleName = v
            end
            currentModuleNameParts = string.split(currentModuleName, ".")
        end
        table.remove(currentModuleNameParts, #currentModuleNameParts)
    end

    return require(moduleFullName)
end

local cc = _G.cc or require('cc')
local AbstractNode = import('.sdk.AbstractNode')

local Node = {}
Node.__index = Node
setmetatable(Node, AbstractNode)

function Node:new(node, screenWidth, screenHeight)
    local n = {}
    setmetatable(n, Node)
    n.node = node
    n.screenWidth = screenWidth
    n.screenHeight = screenHeight
    return n
end


function Node:getParent()
    local parent = self.node:getParent()
    if parent == nil then
        return nil
    end
    return self:new(parent, self.screenWidth, self.screenHeight)
end

function Node:getChildren()
    local children = {}
    for _, child in ipairs(self.node:getChildren()) do
        table.insert(children, self:new(child, self.screenWidth, self.screenHeight))
    end
    return children
end

function Node:getAvailableAttributeNames()
    local ret = {
        'text',
        'touchable',
        'enabled',
        'tag',
        'desc',
        'rotation',
        'rotation3D',
        'skew',
    }
    for _, name in ipairs(AbstractNode.getAvailableAttributeNames(self)) do
        table.insert(ret, name)
    end
    return ret
end

function Node:getAttr(attrName)
    if attrName == 'visible' then
        local visible = self.node:isVisible()
        if not visible then
            return false
        end

        -- if the node is visible, check its parent's visibility
        local parent = self.node:getParent()
        while parent do
            local parentVisible = parent:isVisible()
            if not parentVisible then
                return false
            end
            parent = parent:getParent()
        end
        return true

    elseif attrName == 'name' then
        local name = self.node:getName()
        if name == '' then
            name = self.node:getDescription()
        end
        return name

    elseif attrName == 'text' then
        -- auto strip
        if self.node.getString then
            return self.node:getString():match("^%s*(.-)%s*$")    -- for Label
        elseif self.node.getStringValue then
            return self.node:getStringValue():match("^%s*(.-)%s*$")  -- for TextField
        elseif self.node.getTitleText then
            return self.node:getTitleText():match("^%s*(.-)%s*$")  -- for Button
        end
        return nil

    elseif attrName == 'type' then
        local nodeType = tolua.type(self.node)
        nodeType = nodeType:gsub("^ccui%.", '')
        nodeType = nodeType:gsub("^cc%.", '')
        return nodeType

    elseif attrName == 'pos' then
        -- 转换成归一化坐标系，原点左上角
        local pos = self.node:convertToWorldSpaceAR(cc.p(0, 0))
        pos.x = pos.x / self.screenWidth
        pos.y = pos.y / self.screenHeight
        pos.y = 1 - pos.y
        return {pos.x, pos.y}

    elseif attrName == 'size' then
        -- 转换成归一化坐标系
        local size = self.node:getContentSize()
        -- 有些版本的engine对于某类特殊节点会没有这个值，所以要判断
        if size ~= nil then
            size.width = size.width / self.screenWidth
            size.height = size.height / self.screenHeight
            return {size.width, size.height}
        end

    elseif attrName == 'scale' then
        return {self.node:getScaleX(), self.node:getScaleY()}

    elseif attrName == 'anchorPoint' then
        local anchor = self.node:getAnchorPoint()
        anchor.y = 1 - anchor.y
        return {anchor.x, anchor.y}

    elseif attrName == 'zOrders' then
        local zOrders = {
            global = self.node:getGlobalZOrder(),
            ['local'] = self.node:getLocalZOrder(),
        }
        return zOrders

    elseif attrName == 'touchable' then
        if self.node.isTouchEnabled then
            return self.node:isTouchEnabled()
        end
        return nil

    elseif attrName == 'tag' then
        return self.node:getTag()

    elseif attrName == 'enabled' then
        if self.node.isEnabled then
            return self.node:isEnabled()
        end
        return nil

    elseif attrName == 'desc' then
        return self.node:getDescription()

    elseif attrName == 'rotation' then
        local rotationX, rotationY
        if self.node.getRotationSkewX ~= nil and self.node.getRotationSkewY ~= nil then
            rotationX, rotationY = self.node:getRotationSkewX(), self.node:getRotationSkewY()
        end
        return rotationX or rotationY

    elseif attrName == 'rotation3D' then
        local rotationX, rotationY
        if self.node.getRotationSkewX ~= nil and self.node.getRotationSkewY ~= nil then
            rotationX, rotationY = self.node:getRotationSkewX(), self.node:getRotationSkewY()
        end
        if rotationX == rotationY and self.node.getRotation3D then
            return self.node:getRotation3D()
        end
        return nil

    elseif attrName == 'skew' then
        if self.node.getSkewX and self.node.getSkewY then
            return {self.node:getSkewX(), self.node:getSkewY()}
        end
        return nil

    end

    return AbstractNode.getAttr(self, attrName)
end

function Node:setAttr(attrName, val)
    if attrName == 'text' then
        if self.node.setString then
            self.node:setString(val)
            return true
        elseif self.node.setText then
            self.node:setText(val)
            return true
        end
    end
    return AbstractNode.setAttr(self, attrName, val)
end


return Node